File: stress/metrics/ist.py
from __future__ import annotations

from dataclasses import dataclass
from typing import List, Optional

from OCRB.measure.events import Event, EventType, FailureClass


@dataclass(frozen=True)
class ISTResult:
    ist: Optional[float]                 # None = N/A
    isolation_duration_declared: Optional[float]
    survival_time_observed: Optional[float]
    na_reason: Optional[str] = None


def compute_ist(
    events: List[Event],
    isolation_duration_declared: Optional[float],
) -> ISTResult:
    """
    BP-3 — Isolation Survival Time (IST)

    Normalized definition:
      IST = clamp( survival_time / isolation_duration_declared, 0, 1 )

    Measurement rules (behavioral):
      - Isolation starts at ISOLATION_START.
      - Isolation ends when:
          a) ISOLATION_END occurs (survived isolation), OR
          b) an IRREVERSIBLE failure occurs while isolated, OR
          c) RUN_END occurs while isolated.

    Edge cases:
      - If isolation_duration_declared is missing or <= 0 -> IST is N/A.
      - If no isolation_start event exists -> IST is N/A (not applicable).
    """
    if isolation_duration_declared is None or isolation_duration_declared <= 0:
        return ISTResult(
            ist=None,
            isolation_duration_declared=isolation_duration_declared,
            survival_time_observed=None,
            na_reason="Missing or invalid declared isolation duration.",
        )

    # Find isolation start
    iso_start: Optional[float] = None
    for e in events:
        if e.type == EventType.ISOLATION_START:
            iso_start = e.t_utc
            break

    if iso_start is None:
        return ISTResult(
            ist=None,
            isolation_duration_declared=isolation_duration_declared,
            survival_time_observed=None,
            na_reason="No isolation_start event observed (IST not applicable).",
        )

    # Track whether we're currently isolated
    isolated = True

    # Find first terminating event after iso_start
    iso_end_time: Optional[float] = None

    for e in events:
        if e.t_utc < iso_start:
            continue

        if isolated and e.type == EventType.ISOLATION_END:
            iso_end_time = e.t_utc
            isolated = False
            break

        if isolated and e.type == EventType.FAILURE and e.failure_class == FailureClass.IRREVERSIBLE:
            iso_end_time = e.t_utc
            break

        if isolated and e.type == EventType.RUN_END:
            iso_end_time = e.t_utc
            break

    if iso_end_time is None:
        # No terminating event observed; assume survived until last event timestamp
        iso_end_time = max(e.t_utc for e in events)

    survival_time = max(0.0, iso_end_time - iso_start)
    ist = survival_time / float(isolation_duration_declared)
    ist = max(0.0, min(1.0, ist))

    return ISTResult(
        ist=ist,
        isolation_duration_declared=float(isolation_duration_declared),
        survival_time_observed=survival_time,
        na_reason=None,
    )

File: stress-ref/src/metrics/ist.rs
use crate::types::events::{Event, EventType, FailureClass};

#[derive(Debug, Clone)]
pub struct IstResult {
    pub ist: Option<f64>,
    pub isolation_duration_declared: Option<f64>,
    pub survival_time_observed: Option<f64>,
    pub na_reason: Option<String>,
}

/// BP-3: Isolation Survival Time. IST = clamp(survival_time / declared_duration, 0, 1).
pub fn compute_ist(events: &[Event], isolation_duration_declared: Option<f64>) -> IstResult {
    let declared = match isolation_duration_declared {
        Some(d) if d > 0.0 => d,
        _ => {
            return IstResult {
                ist: None,
                isolation_duration_declared,
                survival_time_observed: None,
                na_reason: Some("Missing or invalid declared isolation duration".to_string()),
            }
        }
    };

    let iso_start = events
        .iter()
        .find(|e| e.event_type == EventType::IsolationStart)
        .map(|e| e.t_utc);

    let iso_start = match iso_start {
        Some(t) => t,
        None => {
            return IstResult {
                ist: None,
                isolation_duration_declared: Some(declared),
                survival_time_observed: None,
                na_reason: Some("No isolation_start event observed".to_string()),
            }
        }
    };

    let mut iso_end_time: Option<f64> = None;
    for e in events {
        if e.t_utc < iso_start {
            continue;
        }
        match e.event_type {
            EventType::IsolationEnd => {
                iso_end_time = Some(e.t_utc);
                break;
            }
            EventType::Failure if e.failure_class == Some(FailureClass::Irreversible) => {
                iso_end_time = Some(e.t_utc);
                break;
            }
            EventType::RunEnd => {
                iso_end_time = Some(e.t_utc);
                break;
            }
            _ => {}
        }
    }

    let end_time = match iso_end_time {
        Some(t) => t,
        None => {
            // No terminating event found — check if there are any events after iso_start
            let max_t = events.iter().map(|e| e.t_utc).fold(f64::NEG_INFINITY, f64::max);
            if max_t <= iso_start {
                // No observable events after isolation started — IST is undefined
                return IstResult {
                    ist: None,
                    isolation_duration_declared: Some(declared),
                    survival_time_observed: None,
                    na_reason: Some("No events observed after isolation start".to_string()),
                };
            }
            max_t
        }
    };

    let survival_time = (end_time - iso_start).max(0.0);
    let ist = (survival_time / declared).clamp(0.0, 1.0);

    IstResult {
        ist: Some(ist),
        isolation_duration_declared: Some(declared),
        survival_time_observed: Some(survival_time),
        na_reason: None,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::events::EventLog;

    #[test]
    fn full_survival() {
        let mut log = EventLog::new("t", "W1-A");
        log.emit_at(EventType::RunStart, 1000.0);
        log.emit_at(EventType::IsolationStart, 1010.0);
        log.emit_at(EventType::IsolationEnd, 1070.0);
        log.emit_at(EventType::RunEnd, 1080.0);

        let result = compute_ist(log.events(), Some(60.0));
        assert_eq!(result.ist, Some(1.0));
        assert_eq!(result.survival_time_observed, Some(60.0));
    }

    #[test]
    fn partial_survival() {
        let mut log = EventLog::new("t", "W1-A");
        log.emit_at(EventType::RunStart, 1000.0);
        log.emit_at(EventType::IsolationStart, 1010.0);
        // Irreversible failure at 1040 — only 30s of 60s survived
        let e = log.emit_at(EventType::Failure, 1040.0);
        e.failure_class = Some(FailureClass::Irreversible);
        log.emit_at(EventType::RunEnd, 1080.0);

        let result = compute_ist(log.events(), Some(60.0));
        assert_eq!(result.ist, Some(0.5));
    }

    #[test]
    fn no_isolation_na() {
        let mut log = EventLog::new("t", "W1-A");
        log.emit_at(EventType::RunStart, 1000.0);
        log.emit_at(EventType::RunEnd, 1080.0);

        let result = compute_ist(log.events(), Some(60.0));
        assert!(result.ist.is_none());
    }

    #[test]
    fn no_declared_duration_na() {
        let mut log = EventLog::new("t", "W1-A");
        log.emit_at(EventType::RunStart, 1000.0);
        log.emit_at(EventType::IsolationStart, 1010.0);
        log.emit_at(EventType::RunEnd, 1080.0);

        let result = compute_ist(log.events(), None);
        assert!(result.ist.is_none());
    }
}
